aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotSslUtility.java

```diff
@@ -15,30 +15,62 @@
 
 package com.amazonaws.mobileconnectors.iot;
 
+import com.amazonaws.logging.Log;
+import com.amazonaws.logging.LogFactory;
+
 import java.security.KeyManagementException;
 import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
 import java.security.SecureRandom;
+import java.security.Security;
 import java.security.UnrecoverableKeyException;
 
 import javax.net.ssl.KeyManager;
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLSocketFactory;
+import org.conscrypt.OpenSSLProvider;
 
 /**
  * AWSIoTSslUtility Utility class for creating an SSL Socket Factory from
  * certs/keys in keystore.
  */
 final class AWSIotSslUtility {
 
+    protected static final String[] ALPN_EXTENSION = {"x-amzn-mqtt-ca"};
+    private static final String TLS_V_1_2 = "TLSv1.2";
+    protected static int portNumber = 8883;
+    private static final Log LOGGER = LogFactory.getLog(AWSIotSslUtility.class);
+
     /**
      * Utility class.
      */
     private AWSIotSslUtility() {
     }
 
+    /**
+     * Creates a socket factory given a keystore and a portNumber.
+     *
+     * @param keyStore keystore containing a certificate and private key for
+     *            used in creating a secured socket.
+     * @param portNum Port number used for connecting to Iot
+     * @return a socket factory for use in creating a secured socket.
+     * @throws NoSuchAlgorithmException when TLS 1.2 is not available.
+     * @throws UnrecoverableKeyException when the private key cannot be
+     *             recovered. Ususally a bad keystore password.
+     * @throws KeyStoreException when keystore cannot be created.
+     * @throws KeyManagementException when SSL context cannot be created by key
+     *             manager.
+     */
+    protected synchronized static SSLSocketFactory getSocketFactoryWithKeyStore(KeyStore keyStore, int portNum)
```

Comment:
The current approach was influenced by reluctance to modify the apparently public method since this wasn't initially planned to be a breaking change but that said I do agree that a utility method should not have variable state that affects other methods. I have retained just `getSocketFactoryWithKeyStore(KeyStore keyStore, int portNum)` method to avoid having to save state.